private/pkg/manifest/manifest.go

```diff
@@ -0,0 +1,269 @@
+	"golang.org/x/crypto/sha3"
+)
+
+var errPartial error = errors.New("partial record")
+
+const shake256Name = "shake256"
+
+type Digest struct {
+	dtype  string
+	digest []byte
+	hexstr string
+}
+
+func NewDigestFromBytes(dtype string, digest []byte) *Digest {
+	return &Digest{
+		dtype:  dtype,
+		digest: digest,
+		hexstr: hex.EncodeToString(digest),
+	}
+}
+
+func NewDigestFromHex(dtype string, hexstr string) (*Digest, error) {
+	digest, err := hex.DecodeString(hexstr)
+	if err != nil {
+		return nil, err
+	}
+	return NewDigestFromBytes(dtype, digest), nil
+}
+
+func NewDigestFromString(typedDigest string) (*Digest, error) {
+	hashfunc, digestStr, found := strings.Cut(typedDigest, ":")
+	if !found {
+		return nil, errors.New("malformed digest string")
+	}
+	return NewDigestFromHex(hashfunc, digestStr)
+}
+
+func (d *Digest) String() string {
+	return d.Type() + ":" + d.Hex()
+}
+
+func (d *Digest) Type() string {
+	return d.dtype
+}
+
+func (d *Digest) Bytes() []byte {
+	return d.digest
+}
+
+func (d *Digest) Hex() string {
+	return d.hexstr
+}
+
+// ManifestError occurs when a manifest is malformed.
+type ManifestError struct {
+	lineno int
+	msg    string
+}
+
+var _ error = (*ManifestError)(nil)
+
+func newManifestError(lineno int, msg string) *ManifestError {
+	return &ManifestError{
+		lineno: lineno,
+		msg:    msg,
+	}
+}
+
+func (e *ManifestError) Error() string {
+	return fmt.Sprintf("invalid manifest: %d: %s", e.lineno, e.msg)
+}
+
+// Manifest represents a list of paths and their digests.
+type Manifest struct {
+	paths   map[string]*Digest
+	digests map[string]string
+	hash    sha3.ShakeHash
+}
+
+var _ encoding.TextMarshaler = (*Manifest)(nil)
+var _ encoding.TextUnmarshaler = (*Manifest)(nil)
+
+// NewManifest creates an empty manifest.
+func NewManifest() *Manifest {
+	return &Manifest{
+		paths:   make(map[string]*Digest),
+		digests: make(map[string]string),
+		hash:    sha3.NewShake256(),
+	}
+}
+
+func splitManifest(data []byte, atEOF bool) (int, []byte, error) {
+	// Return a line without LF.
+	if i := bytes.IndexByte(data, '\n'); i >= 0 {
+		return i + 1, data[0:i], nil
+	}
+
+	// EOF occurred with a partial line.
+	if atEOF && len(data) != 0 {
+		return 0, nil, errPartial
+	}
+
+	return 0, nil, nil
+}
+
+// NewManifestFromReader builds a manifest an encoded manifest reader.
+func NewManifestFromReader(manifest io.Reader) (*Manifest, error) {
+	m := NewManifest()
+	scanner := bufio.NewScanner(manifest)
+	scanner.Split(splitManifest)
+	lineno := 0
+	for scanner.Scan() {
+		lineno++
+		hash, path, found := strings.Cut(scanner.Text(), "  ")
+		if !found {
+			return nil, newManifestError(lineno, "invalid entry")
+		}
+		digest, err := NewDigestFromString(hash)
+		if err != nil {
+			return nil, multierr.Append(
+				newManifestError(lineno, "invalid hash"),
+				err,
+			)
+		}
+		if digest.Type() != shake256Name {
+			return nil, newManifestError(lineno, "unknown hash")
+		}
+		if len(digest.Bytes()) != 64 {
```

Comment:
nit: might be good to make this a constant too (if it is meant to live alongside the expected digest name.

Reply:
I can make it a constant. Really there's a distinction of a hasher and a hash constructor that defines such parameters, but I don't want to boil that ocean until we really want more than one hash type. :)